### PR TITLE
[14.0][IMP] base_comment_template: Include changes from 13.0

### PR DESCRIPTION
**Pull request by victoralmau**
Include changes from 13.0 to standardize both versions:

- Migration to v13: https://github.com/OCA/reporting-engine/pull/481
- Add function to get_comment_template records: https://github.com/OCA/reporting-engine/pull/497
- Fix tests: test_company_general_template: https://github.com/OCA/reporting-engine/pull/499
- Fix Change domain in rule according to company: https://github.com/OCA/reporting-engine/pull/501

Locked by:
- [x] https://github.com/OCA/reporting-engine/pull/504

Please @CarlosRoca13  and @ernestotejeda can you review it?

@Tecnativa TT29138

### CODE REVIEW

**Review by CarlosRoca13 — APPROVED** [review]

**Review by chienandalu — COMMENTED** [review]
Is it very hard to cherry-pick the original commits that include these changes?

**Review by chienandalu — APPROVED** [review]

File: base_comment_template/README.rst
Comment (ernestotejeda) [inline base_comment_template/README.rst]: It looks like you generated the readme for branch 13.0

File: base_comment_template/__manifest__.py
Comment (ernestotejeda) [inline base_comment_template/__manifest__.py]:
```suggestion
    "summary": "Add conditional mako template to any report"
```

File: base_comment_template/tests/fake_models.py
Comment (ernestotejeda) [inline base_comment_template/tests/fake_models.py:12]: this dockstring should be kept?

**Review by ernestotejeda — APPROVED** [review]
LGTM.
Just a few comments.

File: base_comment_template/README.rst
Comment (victoralmau) [reply base_comment_template/README.rst]: Although bot will regenerated it, i have changed.

File: base_comment_template/tests/fake_models.py
Comment (victoralmau) [reply base_comment_template/tests/fake_models.py:12]: It's exists now in 13.0, IMO kep according commits.

File: base_comment_template/README.rst
Comment (pedrobaeza) [reply base_comment_template/README.rst]: This has not been changed at the end

File: base_comment_template/models/__init__.py
Comment (pedrobaeza) [inline base_comment_template/models/__init__.py:2]: I don't see the removal of this file.

File: base_comment_template/models/res_partner.py
Comment (pedrobaeza) [inline base_comment_template/models/res_partner.py:11]: You are losing here the `company_dependent` property, which is important. Migration scripts are needed for converting data from many2one to many2many (and with property) if we change to m2m, but I don't know the use case for requiring a m2m. Where that was changed?

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
@petrus-v your opinion is welcome here as well, as you were the porter.

File: base_comment_template/models/res_partner.py
Comment (pedrobaeza) [reply base_comment_template/models/res_partner.py:11]: Commented in https://github.com/OCA/reporting-engine/pull/481#issuecomment-821799830, it's OK to change m2m + multi-company security, but migration scripts should be provided from those on 14.0.1.0.0 to go to this approach, and not fail when migrating from 13.0.1.0.0 to 14.0.1.0.0.

File: base_comment_template/models/base_comment_template.py
Comment (petrus-v) [inline base_comment_template/models/base_comment_template.py:12]: I'm wondering suggested way to move from previous version to this new version?

It would be nice to provide some compatibility with deprecated method that could be remove later (in the next version ?), or at least raise an explicit message that explain to developer how to manage that change.

For instance `get_value` is in used in other modules likes this OCA module [account_comment_template](https://github.com/OCA/account-invoice-reporting/blob/14.0/account_comment_template/models/account_move.py#L32)

File: base_comment_template/tests/test_base_comment_template.py
Comment (petrus-v) [inline base_comment_template/tests/test_base_comment_template.py:27]: While migration I've had that test to improve the coverage, here you get a nice coverage regarding that code but could be nice to ensure lang param is properly forward by using a different language and asserting expected result, I haven't seen such test case kept.

File: base_comment_template/README.rst
Comment (petrus-v) [inline base_comment_template/README.rst:80]: Does my works has no value? :sweat_smile:

File: base_comment_template/tests/test_base_comment_template.py
Comment (petrus-v) [inline base_comment_template/tests/test_base_comment_template.py]: I'm not sure, should we use decorator instead ?

```suggestion
from odoo.tests import tagged


@tagged("post_install", "-at_install")
class TestCommentTemplate(common.SavepointCase):
```

(not sure you should apply this suggestion import wont be at the right place)

File: base_comment_template/models/ir_model.py
Comment (petrus-v) [inline base_comment_template/models/ir_model.py:24]: I've seen in unit test that a res.users fake model is created, in suppose if we add a test that instantiate that model this code would be called, isn't it ?

also I'm not sure what that code is supposed to do, it's look like a bit too magic I wondering if a mixin would do the trick ?

**Review by petrus-v — CHANGES_REQUESTED** [review]
It would be really lovely to provide a smoothy transition for those (likes me) who use that base module from different place in different other modules !

File: base_comment_template/models/res_partner.py
Comment (victoralmau) [reply base_comment_template/models/res_partner.py:10]: OK but first, it's important to apply m2m in 13.0 in other PR, right?

**Comment by victoralmau** [comment]
According to https://github.com/OCA/reporting-engine/pull/504 it's ready to review now, right?
I'm going to create another PR for `account_comment_template` in 14.0 according to these changes.

File: base_comment_template/models/base_comment_template.py
Comment (victoralmau) [reply base_comment_template/models/base_comment_template.py:12]: Now, it's not necessary.

File: base_comment_template/tests/test_base_comment_template.py
Comment (victoralmau) [reply base_comment_template/tests/test_base_comment_template.py:27]: Now, it's not necessary.

File: base_comment_template/models/base_comment_template.py
Comment (petrus-v) [reply base_comment_template/models/base_comment_template.py:12]: Could you explain why or how ? I'm not sure to understand what's is not necessary, I've some modules (not OCA modules) that's depends on this one.

Comment (victoralmau) [reply base_comment_template/models/base_comment_template.py:12]: IMO, these examples: https://github.com/OCA/account-invoice-reporting/pull/174 (13.0) and https://github.com/OCA/account-invoice-reporting/pull/194 (14.0) will help you understand how other addons (OCA or custom addons) `comment.template` will use and how to display them in a report.

File: base_comment_template/README.rst
Comment (pedrobaeza) [reply base_comment_template/README.rst]: Please regenerate this for not having this double diff

File: base_comment_template/migrations/13.0.1.0.0/pre-migration.py
Comment (pedrobaeza) [inline base_comment_template/migrations/13.0.1.0.0/pre-migration.py]: Remove this, but add the corresponding script for this version, taking into account that you can come from 13.0.2.0.0.

File: base_comment_template/migrations/13.0.2.0.0/pre-migration.py
Comment (pedrobaeza) [inline base_comment_template/migrations/13.0.2.0.0/pre-migration.py]: The same

File: base_comment_template/tests/test_base_comment_template.py
Comment (pedrobaeza) [reply base_comment_template/tests/test_base_comment_template.py]: This shouldn't be put at all. No need to execute at the end.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Attend also the rest of the comments

File: base_comment_template/models/ir_model.py
Comment (victoralmau) [reply base_comment_template/models/ir_model.py:24]: Yest, exactly.
These addon need to test if some model inherit comment template and everything work fine. 
We need add some code to ir_model because only with inherit comment.template we can select model in the templates.
Maybe some documentation about fake_models https://pypi.org/project/odoo-test-helper/ help you.

**Comment by victoralmau** [comment]
It's ready to review.
@petrus-v can you re-review?

File: base_comment_template/migrations/13.0.1.0.0/pre-migration.py
Comment (pedrobaeza) [inline base_comment_template/migrations/13.0.1.0.0/pre-migration.py]: These folders from old version must be removed.

File: base_comment_template/__manifest__.py
Comment (pedrobaeza) [inline base_comment_template/__manifest__.py]: Change this to `14.0.2.0.0`

File: base_comment_template/migrations/14.0.1.0.0/pre-migration.py
Comment (pedrobaeza) [inline base_comment_template/migrations/14.0.1.0.0/pre-migration.py]: This should be executed only if coming from 14.0.1.0.0 to 14.0.2.0.0, so change the folder name, and we need to check if we are coming from 13.0.x or from 14.0.1.0.0, so look for a data condition that is only in 14.0.1.0.0.

It should be basically the same as what we have done from 13.0.1.0.0 to 13.0.2.0.0, that is tested, so remove this not elegant comment and make sure it works.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by pedrobaeza** [comment]
@petrus-v we don't want to break your current module status, but there were also a jump from 12 to 14 in this migration, and with a design that is in the middle of the old and the new one, so we need to unblock this transition (that we are trying to be smooth through migration script). Is there any doubt you have with the changes?

**Comment by petrus-v** [comment]
@pedrobaeza 

> @petrus-v we don't want to break your current module status, but there were also a jump from 12 to 14 in this migration, and with a design that is in the middle of the old and the new one, so we need to unblock this transition (that we are trying to be smooth through migration script). Is there any doubt you have with the changes?

I haven't deep into real impact in my customer context, I've also [suggested change in an other PR](https://github.com/OCA/reporting-engine/pull/495) to be able to use jinja template that I use, but anyway I understand your point go ahead I'll adapt all my code once this is merged.

I'm pretty confident as I've tools and tests to prevent myself to deploy breaking change in production and I'll be able to freeze things in the mean time to adapt code or fix before deploy the last version once I'll be happy.

Thanks again for all your comments, was very appreciated.

**Comment by victoralmau** [comment]
Thanks @pedrobaeza, changes done.

**Review by joao-p-marques — APPROVED** [review]

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**
